projects/admin/src/app/menu/menu-switch-library/service/library-switch-menu.service.ts

```diff
@@ -0,0 +1,217 @@
+ * the Free Software Foundation, version 3 of the License.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { Injectable } from '@angular/core';
+import { TranslateService } from '@ngx-translate/core';
+import { MenuFactory, MenuItem, Record, RecordEvent, RecordService } from '@rero/ng-core';
+import { User } from '@rero/shared';
+import { map } from 'rxjs/operators';
+import { LibraryService } from '../../../service/library.service';
+import { LibrarySwitchService } from './library-switch.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class LibrarySwitchMenuService {
+
+  /** User */
+  private _user: User;
+
+  /** Menu */
+  private _menu: any;
+
+  /** Current Library */
+  private _currentLibrary: MenuItem;
+
+  /** Available libraries */
+  private _librariesMenu = {};
+
+  /** menu is visible */
+  get visible() {
+    return this._user.hasRole('system_librarian')
+      ? true
+      : this._user.libraries.length > 1;
+  }
+
+  /** Get menu */
+  get menu() {
+    return this._menu;
+  }
+
+  /**
+   * Constructor
+   * @param _librarySwitchService - LibrarySwitchService
+   * @param _libraryService - LibraryService
+   * @param _recordService - RecordService
+   * @param _translateService - TranslateService
+   */
+  constructor(
+    private _librarySwitchService: LibrarySwitchService,
+    private _libraryService: LibraryService,
+    private _recordService: RecordService,
+    private _translateService: TranslateService
+  ) { }
+
+  /** Init */
+  init() {
+    this._initObservable();
+  }
+
+  /** Generate */
+  private _generate() {
+    if (this._user.hasRole('system_librarian')) {
+      this._systemLibrarianLibraries();
+    } else if (this._user.libraries.length > 1) {
+      this._librarianLibraries();
+    }
+  }
+
+  /** System librarian libraries */
+  private _systemLibrarianLibraries() {
+    this._executeQuery(
+      this._libraryService.findAllOrderBy$('name')
+    ).subscribe((libraries: any) => {
+      this._generateMenu(libraries);
+    });
+  }
+
+  /** Librarian libraries */
+  private _librarianLibraries() {
+    const librariesPid = this._user.libraries.map(library => {
+      return library.pid;
+    });
+    this._executeQuery(
+      this._libraryService.findByLibrariesPidAndOrderBy$(librariesPid, 'name')
+    ).subscribe((libraries: any) => {
+      this._generateMenu(libraries);
+    });
+  }
+
+  /**
+   * Execute query with an Observable
+   * @param query - Observable
+   * @return Observable
+   */
+  private _executeQuery(query: any) {
```

Comment:
This method can have a more explicit name, for example which kind of query.

Reply:
done